beacon-chain/sync/initial-sync/blocks_fetcher.go

```diff
@@ -640,70 +598,557 @@ func (f *blocksFetcher) fetchBlobsFromPeer(ctx context.Context, bwb []blocks2.Bl
+			log.Warning("No peers available to retrieve missing data columns, retrying in 5 seconds")
+			time.Sleep(5 * time.Second)
 			continue
 		}
-		return robs, err
+
+		// Get the first slot for which we should retrieve data columns.
+		startSlot := bwb[firstIndex].Block.Block().Slot()
+
+		// Build the request.
+		request := &p2ppb.DataColumnSidecarsByRangeRequest{
+			StartSlot: startSlot,
+			Count:     blocksCount,
+			Columns:   sortedSliceFromMap(missingDataColumns),
+		}
+
+		// Get all the blocks and data columns we should retrieve.
+		blockFromRoot := blockFromRoot(bwb[firstIndex : lastIndex+1])
+
+		// Iterate request over all peers, and exit as soon as at least one data column is retrieved.
+		roDataColumns, peer, err := f.requestDataColumnsFromPeers(ctx, request, peers)
+		if err != nil {
+			return errors.Wrap(err, "request data columns from peers")
+		}
+
+		// Process the retrieved data columns.
+		processRetrievedDataColumns(roDataColumns, blockFromRoot, indicesFromRoot, missingColumnsFromRoot, bwb)
+
+		if len(missingColumnsFromRoot) > 0 {
+			for root, columns := range missingColumnsFromRoot {
+				log.WithFields(logrus.Fields{
+					"peer":    peer,
+					"root":    fmt.Sprintf("%#x", root),
+					"slot":    blockFromRoot[root].Block().Slot(),
+					"columns": columns,
+				}).Debug("Peer did not correctly return data columns")
+			}
+		}
 	}
-	return nil, errNoPeersAvailable
+
+	return nil
+}
+
+// fetchDataColumnsFromPeers looks at the blocks in `bwb` and retrieves all
+// data columns for with the block has blob commitments, and for which our store is missing data columns
+// we should custody.
+// Preqrequisite: bwb is sorted by slot.
+func (f *blocksFetcher) fetchDataColumnsFromPeers(
+	ctx context.Context,
+	bwb []blocks2.BlockWithROBlobs,
+	peers []peer.ID,
+) ([]blocks2.BlockWithROBlobs, error) {
+	ctx, span := trace.StartSpan(ctx, "initialsync.fetchColumnsFromPeer")
+	defer span.End()
+
+	// Get the current slot.
+	currentSlot := f.clock.CurrentSlot()
+
+	// If there is no data columns before deneb. Early return.
+	if slots.ToEpoch(currentSlot) < params.BeaconConfig().DenebForkEpoch {
+		return bwb, nil
+	}
+
+	// Get the columns we custody.
+	localCustodyColumns, err := f.custodyColumns()
+	if err != nil {
+		return nil, errors.Wrap(err, "custody columns")
+	}
+
+	// Find the first and last block in `bwb` that contain at least one blob commitment.
+	commitmentsExist, firstIndex, lastIndex, err := blocksWithBlobsCommitmentsBoundaries(bwb, currentSlot)
+	if err != nil {
+		return nil, errors.Wrap(err, "blocks with blobs commitments boundaries")
+	}
+
+	// If there is no block with blob commitments, early return.
+	if !commitmentsExist {
+		return bwb, nil
+	}
+
+	// Find the first and last block for which some data columns to retrieve are missing in our store.
+	someColumnsAreMisisng, firstIndex, lastIndex := f.blocksWithMissingDataColumnsBoundaries(bwb, firstIndex, lastIndex, localCustodyColumns)
```

Comment:
We should combine `blocksWithBlobsCommitmentsBoundaries` and `blocksWithMissingDataColumnsBoundaries` into one method, as we are primarily looking for the block range to start requesting columns from .